### PR TITLE
Just a few std.mem changes

### PR DESCRIPTION
**Pull request by BraedonWooding**
Much smaller PR
- Trim allows you to just trim from left/right side (or both)
  - Can you do bitwise in Zig?  It seems to not like it?  Do you have to cast and bitwise?
- Index commands do allow you to not stop at the first but stop at the last one.
  - While we could index backwards in these cases it would obfuscate the code too much, not worth the loss in readability
- endsWith added to line up with startsWith.

### CODE REVIEW

File: std/mem.zig [hunk 1]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
...
 }
 
 /// Linear search for the index of a scalar value inside a slice.
 pub fn indexOfScalar(comptime T: type, slice: []const T, value: T) ?usize {
-    return indexOfScalarPos(T, slice, 0, value);
+    return indexOfScalarPos(T, slice, 0, value, false);
 }
 
-pub fn indexOfScalarPos(comptime T: type, slice: []const T, start_index: usize, value: T) ?usize {
+pub fn indexOfScalarPos(comptime T: type, slice: []const T, start_index: usize, value: T, highest: bool) ?usize {
```
Comment (andrewrk) [inline std/mem.zig]: this should be `indexOfScalarPos` and `lastIndexOfScalarPos` instead of a boolean parameter.

File: std/mem.zig [hunk 2]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
...
     var i: usize = start_index;
+    var out: ?usize = null;
+
     while (i < slice.len) : (i += 1) {
-        if (slice[i] == value)
-            return i;
+        if (slice[i] == value) {
+            out = i;
+            if (!highest) break;
```
Comment (andrewrk) [inline std/mem.zig]: `lastIndexOfScalarPos` should use a better algorithm than this, searching from the end.

File: std/mem.zig [hunk 3]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
...
+    return out;
 }
 
 pub fn indexOfAny(comptime T: type, slice: []const T, values: []const T) ?usize {
-    return indexOfAnyPos(T, slice, 0, values);
+    return indexOfAnyPos(T, slice, 0, values, false);
 }
 
-pub fn indexOfAnyPos(comptime T: type, slice: []const T, start_index: usize, values: []const T) ?usize {
+pub fn indexOfAnyPos(comptime T: type, slice: []const T, start_index: usize, values: []const T, highest: bool) ?usize {
```
Comment (andrewrk) [inline std/mem.zig]: `indexOfAnyPos` and `lastIndexOfAnyPos` without the bool.

File: std/mem.zig [hunk 4]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
...
     var i: usize = start_index;
+    var out: ?usize = null;
+
     while (i < slice.len) : (i += 1) {
         for (values) |value| {
-            if (slice[i] == value)
-                return i;
+            if (slice[i] == value) {
+                out = i;
+                if (!highest) break;
```
Comment (andrewrk) [inline std/mem.zig]: `lastIndexOfAnyPos` should have a better algorithm than this. at worst searching from the end, at best reverse boyer-moore or something like that.

File: std/mem.zig [hunk 5]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
...
         }
     }
-    return null;
+
+    return out;
 }
 
 pub fn indexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize {
-    return indexOfPos(T, haystack, 0, needle);
+    return indexOfPos(T, haystack, 0, needle, false);
```
Comment (andrewrk) [inline std/mem.zig]: see above comments

File: std/mem.zig [hunk 6]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
...
 
+    var out : ?usize = null;
     var i: usize = start_index;
     const end = haystack.len - needle.len;
     while (i <= end) : (i += 1) {
-        if (eql(T, haystack[i .. i + needle.len], needle))
-            return i;
+        if (eql(T, haystack[i .. i + needle.len], needle)) {
+            out = i;
+            if (!highest) break;
```
Comment (andrewrk) [inline std/mem.zig]: see above comments

File: std/mem.zig [hunk 7]
```diff
@@ -355,6 +382,10 @@ pub fn startsWith(comptime T: type, haystack: []const T, needle: []const T) bool
     return if (needle.len > haystack.len) false else eql(T, haystack[0 .. needle.len], needle);
 }
 
+pub fn endsWith(comptime T: type, haystack: []const T, needle: []const T) bool {
```
Comment (andrewrk) [inline std/mem.zig]: can you add a test for this?

File: std/mem.zig [hunk 8]
```diff
@@ -170,72 +170,99 @@ pub fn dupe(allocator: &Allocator, comptime T: type, m: []const T) ![]T {
     return new_buf;
 }
 
+pub const Side = enum { LEFT = 1, RIGHT = 2, BOTH = 3, };
+
 /// Remove values from the beginning and end of a slice.
-pub fn trim(comptime T: type, slice: []const T, values_to_strip: []const T) []const T {
-    var begin: usize = 0;
+pub fn trim(comptime T: type, slice: []const T, values_to_strip: []const T, side: Side) []const T {
```
[the remark this answers is not in the document]
Comment (BraedonWooding) [reply std/mem.zig]: Fair

File: std/mem.zig (hunk 2 above)
Comment (BraedonWooding) [reply std/mem.zig]: Yes as I said in my comment I didn't want to remove readability for that speed bonus but if you want them as separate functions I'll implement the superior algorithm :).

File: std/mem.zig (hunk 1 above)
Comment (BraedonWooding) [reply std/mem.zig]: Fair, will do

File: std/mem.zig (hunk 7 above)
Comment (BraedonWooding) [reply std/mem.zig]: Sure :).  I'll add a test for starts with too.

File: std/mem.zig (hunk 6 above)
Comment (BraedonWooding) [reply std/mem.zig]: Yep, I'll just reply once or twice as to not repeat myself.  I'll clean this up :).

**Comment by BraedonWooding** [comment]
@andrewrk Squashed commits and made requested changes.

**Comment by andrewrk** [comment]
I fixed some bugs and improved the API, and avoided the merge commit

**Pull request closed without merging**